File: client/scala/src/main/scala/ai/verta/repository/Repository.scala
package ai.verta.repository

import ai.verta.swagger.client.ClientSet
import ai.verta.swagger._public.modeldb.versioning.model.VersioningRepository

import scala.concurrent.ExecutionContext
import scala.util.{Failure, Success, Try}

/** ModelDB Repository.
 *  There should not be a need to instantiate this class directly; please use Client's getOrCreateRepository
 */
class Repository(private val clientSet: ClientSet, private val repo: VersioningRepository) {
  /** Get commit by its SHA id
   * @param id SHA ID of the commit
   * @return specified commit
   */
  def getCommitById(id: String)(implicit ec: ExecutionContext): Try[Commit] = {
    clientSet.versioningService.VersioningService_GetCommit2(
      repository_id_repo_id = repo.id.get,
      commit_sha = id
    ).map(r => new Commit(clientSet, this, r.commit.get))
  }

  /** Get commit by specified branch
   *  @param branch branch of commit. If not passed, then use "master" branch
   *  @return specified commit
   */
   def getCommitByBranch(branch: String = "master")(implicit ec: ExecutionContext): Try[Commit] = {
     clientSet.versioningService.VersioningService_GetBranch2(
       branch = branch,
       repository_id_repo_id = repo.id.get
     ).map(r => new Commit(clientSet, this, r.commit.get, Some(branch)))
   }

   /** Get commit by specified tag
    *  @param tag tag of commit.
    *  @return specified commit
    */
   def getCommitByTag(tag: String)(implicit ec: ExecutionContext): Try[Commit] = {
     clientSet.versioningService.VersioningService_GetTag2(
       tag = tag,
       repository_id_repo_id = repo.id.get
     ).map(r => new Commit(clientSet, this, r.commit.get))
   }

   /** Delete a tag from this repository
    *  @param tag tag
    */
    def deleteTag(tag: String)(implicit ec: ExecutionContext): Try[Unit] = {
      clientSet.versioningService.VersioningService_DeleteTag2(
          repository_id_repo_id = repo.id.get,
          tag = tag
      ).map(_ => ())
    }


    override def equals(other: Any) = other match {
      case other: Repository => repo.id.get == other.repo.id.get
      case _ => false
    }

    /** Return the name of the repository */
    def name = repo.name.get

    /** Return the id of the repository */
    def id = repo.id.get
}

File: client/verta/verta/repository/_repository.py
# -*- coding: utf-8 -*-

from __future__ import print_function

from .._protos.public.modeldb.versioning import VersioningService_pb2 as _VersioningService

from .._internal_utils import _utils
from ..visibility import _visibility
from . import _commit

import requests


class Repository(object):
    """
    ModelDB Repository.

    There should not be a need to instantiate this class directly; please use
    :meth:`Client.get_or_create_repository() <verta.Client.get_or_create_repository>`.

    Attributes
    ----------
    id : str
        ID of the Repository.
    name : str
        Name of the Repository.

    """
    def __init__(self, conn, id_):
        self._conn = conn

        self.id = id_

    def __repr__(self):
        return "<Repository \"{}\">".format(self.name)

    @property
    def _endpoint_prefix(self):
        return "{}://{}/api/v1/modeldb/versioning/repositories/{}".format(
            self._conn.scheme,
            self._conn.socket,
            self.id,
        )

    @property
    def name(self):
        response = _utils.make_request("GET", self._endpoint_prefix, self._conn)
        _utils.raise_for_http_error(response)

        response_msg = _utils.json_to_proto(_utils.body_to_json(response),
                                            _VersioningService.GetRepositoryRequest.Response)
        return response_msg.repository.name

    @property
    def workspace(self):
        # TODO: replace with self._msg when refactor to subclass _ModelDBEntity
        msg = self._get_proto_by_id(self._conn, self.id)

        if msg.workspace_id:
            return self._conn.get_workspace_name_from_legacy_id(msg.workspace_id)
        else:
            return self._conn._OSS_DEFAULT_WORKSPACE

    @classmethod
    def _create(cls, conn, name, workspace, public_within_org, visibility):
        visibility, public_within_org = _visibility._Visibility._translate_public_within_org(visibility, public_within_org)

        msg = _VersioningService.Repository()
        msg.name = name
        if (public_within_org
                and workspace is not None  # not user's personal workspace
                and _utils.is_org(workspace, conn)):  # not anyone's personal workspace
            msg.repository_visibility = _VersioningService.RepositoryVisibilityEnum.ORG_SCOPED_PUBLIC
        msg.custom_permission.CopyFrom(visibility._custom_permission)
        msg.visibility = visibility._visibility

        data = _utils.proto_to_json(msg)
        endpoint = "{}://{}/api/v1/modeldb/versioning/workspaces/{}/repositories".format(
            conn.scheme,
            conn.socket,
            workspace,
        )
        response = _utils.make_request("POST", endpoint, conn, json=data)
        _utils.raise_for_http_error(response)

        response_msg = _utils.json_to_proto(_utils.body_to_json(response),
                                            _VersioningService.SetRepository.Response)
        return cls(conn, response_msg.repository.id)

    @classmethod
    def _get_proto_by_id(cls, conn, id):
        Message = _VersioningService.GetRepositoryRequest
        response = conn.make_proto_request("GET",
                                           "/api/v1/modeldb/versioning/repositories/{}".format(id))
        return conn.maybe_proto_response(response, Message.Response).repository

    @classmethod
    def _get_proto_by_name(cls, conn, name, workspace):
        Message = _VersioningService.GetRepositoryRequest
        response = conn.make_proto_request("GET",
                                           "/api/v1/modeldb/versioning/workspaces/{}/repositories/{}".format(workspace, name))
        return conn.maybe_proto_response(response, Message.Response).repository

    @classmethod
    def _get(cls, conn, name=None, workspace=None, id_=None):
        if name and workspace and not id_:
            msg = cls._get_proto_by_name(conn, name, workspace)
        elif not name and not workspace and id_:
            msg = cls._get_proto_by_id(conn, id_)
        else:
            raise RuntimeError("the Client has encountered an error;"
                               " please notify the Verta development team")

        if not msg:
            return None
        return cls(conn, msg.id)

    def get_commit(self, branch=None, tag=None, id=None):
        """
        Returns the Commit with the specified `branch`, `tag`, or `id`.

        If no arguments are passed, ``branch="master"`` is the default.

        Parameters
        ----------
        branch : str, optional
            Branch of the Commit.
        tag : str, optional
            Tag of the Commit.
        id : str, optional
            ID of the Commit.

        Returns
        -------
        :class:`Commit`
            Specified Commit.

        """
        num_args = sum(map(lambda x: x is not None, [tag, id, branch]))
        if num_args > 1:
            raise ValueError("cannot specify more than one of `branch`, `tag`, and `id`")
        if num_args == 0:
            branch = "master"

        if branch is not None:
            msg = _VersioningService.GetBranchRequest()
            endpoint = "{}://{}/api/v1/modeldb/versioning/repositories/{}/branches/{}".format(
                self._conn.scheme,
                self._conn.socket,
                self.id,
                branch,
            )
        elif tag is not None:
            msg = _VersioningService.GetTagRequest()
            endpoint = "{}://{}/api/v1/modeldb/versioning/repositories/{}/tags/{}".format(
                self._conn.scheme,
                self._conn.socket,
                self.id,
                tag,
            )
        elif id is not None:
            msg = _VersioningService.GetCommitRequest()
            endpoint = "{}://{}/api/v1/modeldb/versioning/repositories/{}/commits/{}".format(
                self._conn.scheme,
                self._conn.socket,
                self.id,
                id,
            )
        response = _utils.make_request("GET", endpoint, self._conn)
        _utils.raise_for_http_error(response)

        response_msg = _utils.json_to_proto(_utils.body_to_json(response), msg.Response)
        return _commit.Commit._from_id(self._conn, self, response_msg.commit.commit_sha, branch_name=branch)

    def delete(self):
        """
        Deletes this repository.

        """
        request_url = "{}://{}/api/v1/modeldb/versioning/repositories/{}".format(self._conn.scheme, self._conn.socket, self.id)
        response = requests.delete(request_url, headers=self._conn.auth)
        _utils.raise_for_http_error(response)
